### ISSUE TITLE
Presence of any command line job-option prevents ioengine-specific options in configuration files

### ISSUE THREAD

**Issue by benlwalker** [axboe/fio#408]
The presence of any command line job-option causes fio to treat all ioengine-specific options in configuration files as invalid. For example, take this configuration file:

```
[global]
ioengine=libaio
userspace_reap=1
...

[job0]
...
```

Running fio with a command line like such:

```
fio <my_config.fio> --iodepth=8
```

results in the following error:

```
Bad option <userspace_reap=1>
```

When fio encounters a command line option for a job (parse_cmd_line - init.c line 2504), if a job does not currently exist it creates a new one with default settings. It then loads the ioengine for that job, which on Linux will be psync. Later, the ioengine option may be encountered. If the ioengine option is on the command line, the code to handle it is in init.c line 2532 and it will unload psync and load the specified ioengine.

When later parsing the configuration file, when an unknown option (i.e. an engine-specific option) is encountered, a list is saved and reprocessed after all known options. However, when the ioengine option (a known option) is encountered, it just updates the string name of the ioengine - it doesn't unload the default engine like it does when the option is on the command line. In the loop to process unknown options it does correctly attempt to load the ioengine, but that call is short-circuited because td->io_ops is already set (because a different ioengine is already loaded).

To recap:

1) All options in configuration file works.
2) All options in command line, if ioengine appears before engine-specific options, works.
3) Engine-specific options that appear before the ioengine option will not work. This is minor - most people will put the ioengine first if they're putting everything on the command line.
4) Engine-specific options on the command line will fail if ioengine is specified in the configuration file. This is a real annoyance for users that have a template config file and then vary engine-specific options on the command line.
5) The presence of any job-option on the command line (like --iodepth) breaks all engine-specific options in configuration files. This is the primary issue.

**Comment by sitsofe** [comment]
1, 2 describe ok behaviour. 5 has been fixed by #409. I'd argue that 3 is kind of a non-issue which leaves 4 as the remaining issue...

**Comment by redlicha** [comment]
Commit 800334db17a22029553488b41a5ede8af909c66d breaks engine-specific options for external engines that are configured within a job section. Example:
```
$ git diff engines/skeleton_external.c
diff --git a/engines/skeleton_external.c b/engines/skeleton_external.c
index 56f89f95..920a25be 100644
--- a/engines/skeleton_external.c
+++ b/engines/skeleton_external.c
@@ -123,6 +123,7 @@ static int fio_skeleton_prep(struct thread_data *td, struct io_u *io_u)
  */
 static int fio_skeleton_init(struct thread_data *td)
 {
+       log_info("skeleton: dummy %u\n", ((struct fio_skeleton_options*)td->eo)->dummy);
        return 0;
 }

$ gcc -D_LARGEFILE64_SOURCE -shared -fPIC -ggdb3 -o ../install/skeleton_external.so engines/skeleton_external.c
$ ../install/bin/fio --ioengine=external:../install/skeleton_external.so --size=4k  --thread --name xxx --dummy=1
xxx: (g=0): rw=read, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=engine_name, iodepth=1
fio-3.2-79-gdb37
Starting 1 thread
xxx: Laying out IO file (1 file / 0MiB)
skeleton: dummy 0
[...]
```
Whereas previously engine options specified per job were picked up:
```
$ ../install/bin/fio.without-pr-409 --ioengine=external:../install/skeleton_external.so --size=4k  --thread --name xxx --dummy=1
xxx: (g=0): rw=read, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=engine_name, iodepth=1
fio-3.2-79-gdb37-dirty
Starting 1 thread
xxx: Laying out IO file (1 file / 0MiB)
skeleton: dummy 1
[..]
```

The reason is that for external engines, `thread_options::ioengine` never contains the same name as `ioengine_ops::name`, but either "external" (in case of `ioengine=external:/path/to/dso`) or the path to the DSO (`ioengine=/path/to/dso`). This leads to `free_ioengine` getting invoked, dropping the parsed options from `td->eo`.

**Comment by axboe** [comment]
You are right, that commit introduced a regression. I've pushed out a fix, can you test if current -git works? This is the fix:

http://git.kernel.dk/cgit/fio/commit/?id=3e262e2fc8e73bae1340aef23a78195b5de82387

**Comment by redlicha** [comment]
The above commit restores the desired behaviour:
```
$ ../install/bin/fio --ioengine=external:../install/skeleton_external.so --size=4k  --thread --name xxx --dummy=1
xxx: (g=0): rw=read, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=engine_name, iodepth=1
fio-3.2-82-g4784
Starting 1 thread
xxx: Laying out IO file (1 file / 0MiB)
skeleton: dummy 1
[...]
```
Thanks!

Looking at the code, shouldn't `dlclose(dlhandle)` also be called in the case of `load_engine` returning the already loaded one to decrement the refcount?

**Comment by axboe** [comment]
The manual dlclose() closes the old one, the free_ioengine() that is called right after will dlclose() the new one.

**Comment by redlicha** [comment]
I meant this branch
```
+		if (ops == td->io_ops && dlhandle == td->io_ops_dlhandle)
+			return 0;
```

**Comment by axboe** [comment]
Ah, yes, good catch, we do leak one for this case. I've committed a fix for that now.
